src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,162 @@
+    {"name": "lfric real scalar", "intrinsic": "real",
+     "precision": "r_def", "vars": [], "short_name": "real_scalar"},
+    {"name": "lfric logical scalar", "intrinsic": "boolean",
+     "precision": "l_def", "vars": [], "short_name": "logical_scalar"},
+    {"name": "cell position", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "mesh height", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "operator size", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "ncell_3d"},
+    {"name": "number of dofs", "intrinsic": "integer",
+     "precision": "i_def", "vars": ["fs"], "short_name": "ndf"},
+    {"name": "number of unique dofs", "intrinsic": "integer",
+     "precision": "i_def", "vars": ["fs"], "short_name": "undf"},
+    {"name": "number of faces", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nfaces"},
+    {"name": "number of edges", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nedges"},
+    {"name": "number of qr points in horizontal", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp_h"},
+    {"name": "number of qr points in vertical", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp_v"},
+    {"name": "number of qr points", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp"}]
+
+for info in scalar_datatypes:
+    NAME = "".join(info["name"].title().split())
+    INTRINSIC = info["intrinsic"].upper()
+    PRECISION = info["precision"]
+    ARGS = ["self", "name"] + info["vars"] + ["interface=None"]
+    VARS = ["        self.{0} = {0}".format(var) for var in info["vars"]]
+    exec('''
+class {0}DataType(ScalarType):
+    def __init__(self):
+        super({0}DataType, self).__init__(
+            ScalarType.Intrinsic.{1}, {2})
+    '''.format(NAME, INTRINSIC, PRECISION.upper()))
+    exec('''
+class {0}DataSymbol(DataSymbol):
+    def __init__({1}):
+{2}
+        super({0}DataSymbol, self).__init__(
+            name, {0}DataType(), interface=interface)
+    '''.format(NAME, ", ".join(ARGS), "\n".join(VARS)))
+
+field_datatypes = [
+    {"name": "real field data", "scalar_type": "lfric real scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]},
+    {"name": "integer field data", "scalar_type": "lfric integer scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]},
+    {"name": "logical field data", "scalar_type": "lfric logical scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]}]
+
+# TBD: #xxx Sort out how to deal with multi-dimensional array
+# information. (diff) basis function dimensions change depending on
+# the type of quadrature and we also want to capture the ordering of
+# the datatypes somehow (see operator:dim_info for example).
+array_datatypes = [
```

Comment:
fields are also 'array datatypes' so I'm not sure of the distinction you're making here.

Reply:
It is simply because fields are subclassed as vector fields and having a separate list simplifies this. I have added comments explaining this.